### PR TITLE
GPU Logistic Regression

### PR DESCRIPTION
**Pull request by christopherbunn**

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -149,20 +152,27 @@ if (enable-test)
...
 
   # Add source files according to machine setup
-  if (CUDA_HAVE_GPU)
+  if (CUDA_AND_GPU)
+  message("MAKING GPU TESTS")
     file(GLOB_RECURSE TEST_SRC_FILES RELATIVE
          ${PROJECT_SOURCE_DIR}
          test/cpu_operations_test/*.cc
-         test/util_test/*.cc
+         test/logistic_regression_test/*.cc
```
Comment (yiskylee) [inline cpp/CMakeLists.txt]: Could you move all logistic regression related tests to NICE/cpp/test/model_test?

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -149,20 +152,27 @@ if (enable-test)
...
+         test/gpu_logistic_regression_test/*.cc
+	       test/util_test/*.cc
          test/cpu_solver_test/*.cc
          test/gpu_operations_test/*.cc
          test/gpu_solver_test/*.cc
-         test/model_test/*.cc
+         test/matrix_vector_mult_test/*.cc
```
Comment (yiskylee) [inline cpp/CMakeLists.txt]: And move matrix vector multiplication related tests to NICE/cpp/test/gpu_operations_test

File: cpp/include/cpu_operations.h [hunk 3]
```diff
@@ -52,7 +55,7 @@ class CpuOperations {
   }
 
   /// This is a function that calculates the transpose Vector of the
-  /// input Vector
+  // input Vecto
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: typo and short of a slash

File: cpp/include/cpu_operations.h [hunk 4]
```diff
@@ -90,7 +93,13 @@ class CpuOperations {
   /// This function returns a Matrix of type T
   static Matrix<T> Multiply(const Matrix<T> &a, const Matrix<T> &b) {
     // Matrix-matrix multiplication
-    return a * b;
+    Matrix<T> result;
+    high_resolution_clock::time_point t1 = high_resolution_clock::now();
+    result = a * b;
+    high_resolution_clock::time_point t2 = high_resolution_clock::now();
+    auto duration = duration_cast<microseconds>( t2 - t1 ).count();
+    std::cout << "CPU Time: " << (long)duration << std::endl;
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: CpuOperations should be clean of timing or anything that is not related to just computing the Multiply. You can do timing in unit tests

File: cpp/include/cuda_shared_MV_multiply.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
+// The MIT License (MIT)
```
Comment (yiskylee) [inline cpp/include/cuda_shared_MV_multiply.h]: no capitalized letters are allowed in a file name, you can rename it cuda_matrix_vector_multiply_shared_memory.h

File: cpp/include/gpu_operations.h [hunk 6]
```diff
@@ -134,6 +165,83 @@ class GpuOperations {
...
+
+      high_resolution_clock::time_point t1 = high_resolution_clock::now();
+      // Set up and do cublas matrix multiply
+      GpuMatrixVectorMul(util_->GetBlasHandle(), norm, m, k, &alpha,
+                        d_a, lda, d_x, incx, &beta, d_y, incy);
+
+      // Device sync
+      util_->SyncDev();
+
+      high_resolution_clock::time_point t2 = high_resolution_clock::now();
```
Comment (yiskylee) [inline cpp/include/gpu_operations.h]: Same as before, please do timing stuff in unit tests.

File: cpp/include/logistic_regression.h [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  void Fit(const Matrix<T> &xin, const Vector<T> &y,
+    int iterations, T alpha){
+    Vector<T> gradient;
+    theta.resize(xin.cols() + 1);
+    gradient.resize(theta.rows());
+    theta.setZero();
+    gradient.setZero();
+    for (int i = 0; i < iterations; i++) {
+      Vector<T> Xtheta = (xin * (theta.bottomRows(theta.rows() - 1)));
+      Xtheta = Xtheta.array() + theta(0);
```
Comment (yiskylee) [inline cpp/include/logistic_regression.h]: no capitalized letter in variable names

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+      CUDA_CALL(cudaMemcpy(d_x, h_x, k * sizeof(T),
+          cudaMemcpyHostToDevice));
+
+      CUDA_CALL(cudaMalloc(&d_y, m * sizeof(T)));
+      CUDA_CALL(cudaMemset(d_y, 0, m * sizeof(T)));
+
+      // Launch kernel here
+      //dim3 dimBlock(BLOCK_SIZE *BLOCK_SIZE);
+      //dim3 dimGrid((a.rows() / dimBlock.x) * (a.cols() / dimBlock.y));
+      high_resolution_clock::time_point t1 = high_resolution_clock::now();
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_matrix_vector_multiply.cu]: I know you are testing individual kernels here and it is hard to do it inside unit tests, what you can do is start a branch that does these timing just for your paper, while having a clean branch where timing info is removed so I can merge them into the develop branch.

File: cpp/src/cuda/cuda_shared_MV_multiply.cu [hunk 9]
```diff
@@ -0,0 +1,148 @@
+// The MIT License (MIT)
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_shared_MV_multiply.cu]: file naming violation

File: cpp/test/shared_MV_mult_test/cuda_shared_MV_multiply_test.cc [hunk 10]
```diff
@@ -0,0 +1,171 @@
+// The MIT License (MIT)
```
Comment (yiskylee) [inline cpp/test/shared_MV_mult_test/cuda_shared_MV_multiply_test.cc]: File naming

File: cpp/test/model_test/logistic_regression_benchmark.cc [hunk 11]
```diff
@@ -43,12 +43,16 @@ class Benchmark: public ::testing::Test {
   Nice::Vector<T> expected_vals;
```
Comment (yiskylee) [inline cpp/test/model_test/logistic_regression_benchmark.cc]: Please move your benchmark tests to a directory named performance_test, so this file's path would be something like:

cpp/test/performance_test/logistic_regression_performance

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (christopherbunn) [reply cpp/CMakeLists.txt]: Fixed.

File: cpp/CMakeLists.txt (hunk 2 above)
Comment (christopherbunn) [reply cpp/CMakeLists.txt]: Fixed.

File: cpp/include/cpu_operations.h (hunk 3 above)
Comment (christopherbunn) [reply cpp/include/cpu_operations.h]: Fixed.

File: cpp/include/cpu_operations.h (hunk 4 above)
Comment (christopherbunn) [reply cpp/include/cpu_operations.h]: Fixed.

File: cpp/include/cuda_shared_MV_multiply.h (hunk 5 above)
Comment (christopherbunn) [reply cpp/include/cuda_shared_MV_multiply.h]: Fixed.

File: cpp/include/gpu_operations.h (hunk 6 above)
Comment (christopherbunn) [reply cpp/include/gpu_operations.h]: Fixed.

File: cpp/include/logistic_regression.h (hunk 7 above)
Comment (christopherbunn) [reply cpp/include/logistic_regression.h]: Fixed.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu (hunk 8 above)
Comment (christopherbunn) [reply cpp/src/cuda/cuda_matrix_vector_multiply.cu]: All benchmark and timing functionality will be moved to Benchmark branch.

File: cpp/src/cuda/cuda_shared_MV_multiply.cu (hunk 9 above)
Comment (christopherbunn) [reply cpp/src/cuda/cuda_shared_MV_multiply.cu]: Fixed.

File: cpp/test/shared_MV_mult_test/cuda_shared_MV_multiply_test.cc (hunk 10 above)
Comment (christopherbunn) [reply cpp/test/shared_MV_mult_test/cuda_shared_MV_multiply_test.cc]: Fixed.

**Review by yiskylee — COMMENTED** [review]
Please address the comments and make sure to reply them.

File: cpp/test/model_test/logistic_regression_benchmark.cc (hunk 11 above)
Comment (yiskylee) [reply cpp/test/model_test/logistic_regression_benchmark.cc]: Because we have decided that performance test should be in its own branch, please scratch the last comment, just move everything performance related to your performance branch.

File: cpp/test/logistic_regression_test/gpu_logistic_regression_test.cc [hunk 12]
```diff
@@ -1,148 +0,0 @@
-// The MIT License (MIT)
```
Comment (christopherbunn) [inline cpp/test/logistic_regression_test/gpu_logistic_regression_test.cc]: This test is redundant, the working one is in model_test

File: cpp/test/model_test/logistic_regression_benchmark.cc (hunk 11 above)
Comment (christopherbunn) [reply cpp/test/model_test/logistic_regression_benchmark.cc]: Fixed, see new commit.

File: cpp/CMakeLists.txt [hunk 13]
```diff
@@ -85,6 +85,9 @@ if (CUDA_AND_GPU)
   # Set source files
   file(GLOB_RECURSE SOURCES RELATIVE ${CMAKE_SOURCE_DIR} src/*.cc src/*.cu)
 
+  # Set source files
+  file(GLOB_RECURSE SOURCES RELATIVE ${CMAKE_SOURCE_DIR} src/*.cc src/cuda/*.cu)
```
Comment (yiskylee) [inline cpp/CMakeLists.txt]: Could you try removing this line but add src/cuda/*.cu to line 86, so you have something like the following:
file(GLOB_RECURSE SOURCES RELATIVE ${CMAKE_SOURCE_DIR} src/*.cc src/*.cu src/cuda/*.cu)
I think this should include your .cu files in the build.

File: cpp/CMakeLists.txt [hunk 14]
```diff
@@ -149,20 +152,24 @@ if (enable-test)
...
          ${PROJECT_SOURCE_DIR}
          test/cpu_operations_test/*.cc
-         test/util_test/*.cc
+	       test/util_test/*.cc
          test/cpu_solver_test/*.cc
          test/gpu_operations_test/*.cc
          test/gpu_solver_test/*.cc
-         test/model_test/*.cc
+         test/model_test/gpu*.cc
+         test/model_test/logistic*.cc
```
Comment (yiskylee) [inline cpp/CMakeLists.txt]: Please combine line 165 and 164 as one line:
test/model_test/*.cc

File: cpp/CMakeLists.txt [hunk 15]
```diff
@@ -149,20 +152,24 @@ if (enable-test)
...
          )
   else()
+  message("NO GPU TEST MADE")
     file(GLOB_RECURSE TEST_SRC_FILES RELATIVE
          ${PROJECT_SOURCE_DIR}
          test/cpu_operations_test/*.cc
+	 test/logistic_regression_test/*.cc
```
Comment (yiskylee) [inline cpp/CMakeLists.txt]: Please delete this line as there is no logistic_regression_test any more

File: cpp/include/cpu_operations.h [hunk 16]
```diff
@@ -32,6 +32,9 @@
 #include "Eigen/SVD"
 #include "include/svd_solver.h"
 #include "include/util.h"
+#include <chrono>
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: Please delete chrono and its namespace as performance tests are excluded from this file.

File: cpp/include/gpu_operations.h [hunk 17]
```diff
@@ -34,7 +34,7 @@
 #include <cusolverDn.h>
 #include <unistd.h>
 #include <stdexcept>
-#include <ctime>
+#include <chrono>
```
Comment (yiskylee) [inline cpp/include/gpu_operations.h]: chrono

File: cpp/include/gpu_operations.h [hunk 18]
```diff
@@ -43,6 +43,8 @@
 #include "include/gpu_util.h"
 #include "include/gpu_svd_solver.h"
 
+using namespace std::chrono;
```
Comment (yiskylee) [inline cpp/include/gpu_operations.h]: chrono

File: cpp/include/gpu_operations.h [hunk 19]
```diff
@@ -95,7 +115,16 @@ class GpuOperations {
     return h_c;
   }
 
-  static Matrix<T> Multiply(const Matrix<T> &a, const Matrix<T> &b) {
+  /// This function multiplies an input Matrix and a Matrix
```
Comment (yiskylee) [inline cpp/include/gpu_operations.h]: I might have forgotten the reason, but why is matrix matrix multiply commented out?

File: cpp/include/logistic_regression.h [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+#include <chrono>
+#include "include/matrix.h"
+#include "include/vector.h"
+#include "include/kernel_types.h"
+#include "Eigen/SVD"
+#include "include/svd_solver.h"
+#include "include/util.h"
+
+
+using namespace std::chrono;
```
Comment (yiskylee) [inline cpp/include/logistic_regression.h]: chrono

File: cpp/include/logistic_regression.h [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+    for (int i = 0; i < iterations; i++) {
+      Vector<T> x_theta_mult = (xin * (theta.bottomRows(theta.rows() - 1)));
+      x_theta_mult = x_theta_mult.array() + theta(0);
+      gradient.bottomRows(gradient.rows() - 1) =
+        xin.transpose() * (h(x_theta_mult) - y);
+      gradient(0) = theta.sum();
+      theta = theta - ((alpha/ y.size()) * gradient);
+      predictions = Predict(inputs);
+      predictions = predictions.unaryExpr(std::ptr_fun<T,T>(std::round));
+      //if (((predictions - y).squaredNorm() / predictions.size()) <= .05){
```
Comment (yiskylee) [inline cpp/include/logistic_regression.h]: Please remove the comments here. You can print things out in your performance branch.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+    int row = blockIdx.y * blockDim.y + threadIdx.y;
+    int col = blockIdx.x * blockDim.x + threadIdx.x;
+    float sum = 0.0f;
+    if (row >= x_size || col >= a_rows) return;
+    for (int k = 0; k < x_size; k++) {
+      sum += (d_a[k * a_rows + col] * d_x[row * a_rows + k]);
+    }
+    __syncthreads();
+    d_y[row * a_rows + col] = sum;
+    __syncthreads();
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_matrix_vector_multiply.cu]: No need to syncthreads() here as all threads will exit at this point.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+#include "include/cuda_matrix_vector_multiply.h"
+#include <chrono>
+#define BLOCK_SIZE 16
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_matrix_vector_multiply.cu]: Remove BLOCK_SIZE if it is not used.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+  template <typename T>
+  __global__ void CudaMVKernel(T *d_a, T *d_x, T *d_y, int a_rows, int x_size) {
+    int row = blockIdx.y * blockDim.y + threadIdx.y;
+    int col = blockIdx.x * blockDim.x + threadIdx.x;
+    float sum = 0.0f;
+    if (row >= x_size || col >= a_rows) return;
+    for (int k = 0; k < x_size; k++) {
+      sum += (d_a[k * a_rows + col] * d_x[row * a_rows + k]);
+    }
+    __syncthreads();
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_matrix_vector_multiply.cu]: No need to syncthreads() here because you are not using shared memory.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+namespace Nice {
+
+  template <typename T>
+  __global__ void CudaMVKernel(T *d_a, T *d_x, T *d_y, int a_rows, int x_size) {
+    int row = blockIdx.y * blockDim.y + threadIdx.y;
+    int col = blockIdx.x * blockDim.x + threadIdx.x;
+    float sum = 0.0f;
+    if (row >= x_size || col >= a_rows) return;
+    for (int k = 0; k < x_size; k++) {
+      sum += (d_a[k * a_rows + col] * d_x[row * a_rows + k]);
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_matrix_vector_multiply.cu]: I don't quite get this kernel. The indexing seems odd to me, have you tested this? I don't see related tests in cpp/test/gpu_operations_test/

File: cpp/src/cuda/cuda_matrix_vector_multiply_shared.cu [hunk 26]
```diff
@@ -0,0 +1,123 @@
+// The MIT License (MIT)
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_matrix_vector_multiply_shared.cu:1]: There is this file and cuda_matrix_vector_multiply_shared_memory.cu, should we remove this one?

**Review by yiskylee — COMMENTED** [review]
Please reply the comments that I made before making the changes.

File: cpp/CMakeLists.txt (hunk 13 above)
Comment (christopherbunn) [reply cpp/CMakeLists.txt]: Done. Fixed in next push.

File: cpp/src/cuda/cuda_matrix_vector_multiply_shared.cu (hunk 26 above)
Comment (christopherbunn) [reply cpp/src/cuda/cuda_matrix_vector_multiply_shared.cu:1]: I think for now I want to keep it as a baseline for a non-shared memory implementation. When the logistic_regression code works I can remove this.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu (hunk 22 above)
Comment (christopherbunn) [reply cpp/src/cuda/cuda_matrix_vector_multiply.cu]: Fixed.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu (hunk 24 above)
Comment (christopherbunn) [reply cpp/src/cuda/cuda_matrix_vector_multiply.cu]: Fixed.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu (hunk 25 above)
Comment (christopherbunn) [reply cpp/src/cuda/cuda_matrix_vector_multiply.cu]: I have tested it and it works. I can find the related tests for this.

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu (hunk 23 above)
Comment (christopherbunn) [reply cpp/src/cuda/cuda_matrix_vector_multiply.cu]: Fixed

File: cpp/include/logistic_regression.h (hunk 21 above)
Comment (christopherbunn) [reply cpp/include/logistic_regression.h]: Fixed

File: cpp/include/logistic_regression.h (hunk 20 above)
Comment (christopherbunn) [reply cpp/include/logistic_regression.h]: Fixed

File: cpp/include/gpu_operations.h (hunk 19 above)
Comment (christopherbunn) [reply cpp/include/gpu_operations.h]: I think this was to make sure that matrix vector version of the Eigen function was running rather than the matrix matrix version for testing. It is uncommented in the next push.

File: cpp/include/cpu_operations.h (hunk 16 above)
Comment (christopherbunn) [reply cpp/include/cpu_operations.h]: Fixed

File: cpp/include/gpu_operations.h (hunk 17 above)
Comment (christopherbunn) [reply cpp/include/gpu_operations.h]: Fixed

File: cpp/include/gpu_operations.h (hunk 18 above)
Comment (christopherbunn) [reply cpp/include/gpu_operations.h]: Fixed

File: cpp/CMakeLists.txt (hunk 14 above)
Comment (christopherbunn) [reply cpp/CMakeLists.txt]: This was done to exclude the tests for KDAC. I don't have the model tests

File: cpp/CMakeLists.txt (hunk 15 above)
Comment (christopherbunn) [reply cpp/CMakeLists.txt]: Fixed

**Comment by yiskylee** [comment]
Hi Chris, 
I ran into the following errors when doing make, which os and compiler do you use on your machine?


In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/center_test.cc:28:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/outer_product_test.cc:29:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/frobenius_norm_test.cc:31:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/multiply_test.cc:28:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/standard_deviation_test.cc:28:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/logical_not_test.cc:31:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/rank_test.cc:28:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/gen_kernel_matrix_test.cc:29:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/gen_degree_matrix_test.cc:29:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/determinant_test.cc:33:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
make[2]: *** [CMakeFiles/Nice_test.dir/test/cpu_operations_test/center_test.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/logical_or_test.cc:27:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/dot_product_test.cc:36:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/logical_and_test.cc:34:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/matrix_subtract_test.cc:27:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
make[2]: *** [CMakeFiles/Nice_test.dir/test/cpu_operations_test/frobenius_norm_test.cc.o] Error 1
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/inverse_test.cc:33:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;
                            ^
In file included from /home/xiangyu/Dropbox/git_project/NICE/cpp/test/cpu_operations_test/add_test.cc:32:0:
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:22: error: ‘chrono’ is not a namespace-name
 using namespace std::chrono;
                      ^
/home/xiangyu/Dropbox/git_project/NICE/cpp/include/cpu_operations.h:36:28: error: expected namespace-name before ‘;’ token
 using namespace std::chrono;

**Comment by christopherbunn** [comment]
I fixed the issue in the newest commit, but I am running into this problem during compilation. I think it has to do with kdac? 
image

**Comment by yiskylee** [comment]
Have you run into this chrono problem before? If not, have you run into the KDAC problem when there were no issues related to chrono?

**Comment by christopherbunn** [comment]
The Chrono problem was because I accidentally left it there during the removal of all of the timing functionality in this branch's code. I have run into the kdac issue before when there were no problems related to Chrono. I believe there was a fix upstream you submitted that fixed it but I cannot remember what it was.

**Comment by yiskylee** [comment]
Please remove this line: https://github.com/christopherbunn/NICE/blob/da134aac7ea406ea83c9539732250615b345b4d6/cpp/CMakeLists.txt#L89

The removal of star sign as compared to

https://github.com/christopherbunn/NICE/blob/da134aac7ea406ea83c9539732250615b345b4d6/cpp/CMakeLists.txt#L86

makes the SOURCES an empty variable. You can see that CMake complains about it when you do cmake -Denable-test=ON ..

You have called ADD_LIBRARY for library Nice without any source files. This typically indicates a problem with your CMakeLists.txt file

**Comment by yiskylee** [comment]
After Fixing that, please go to https://github.com/christopherbunn/NICE/blob/da134aac7ea406ea83c9539732250615b345b4d6/cpp/include/kdac_gpu.h#L153 and https://github.com/christopherbunn/NICE/blob/da134aac7ea406ea83c9539732250615b345b4d6/cpp/include/kdac_gpu.h#L157
And change (&(this->gamma_matrix_))(0), (&(this->g_of_w_))(0) to &(this->gamma_matrix_)(0), &(this->g_of_w_)(0) respectively, which is to remove the enclosing parenthesis on &(this->"some variable")

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu [hunk 27]
```diff
@@ -20,17 +20,18 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 #include "include/cuda_matrix_vector_multiply.h"
-
+#define BLOCK_SIZE 7
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_matrix_vector_multiply.cu]: A BLOCK_SIZE of 7 is not a good idea. It should always be multiple of 32 (warp size)

File: cpp/src/cuda/cuda_matrix_vector_multiply.cu [hunk 28]
```diff
@@ -64,9 +66,9 @@ namespace Nice {
       CUDA_CALL(cudaMemset(d_y, 0, m * sizeof(T)));
 
       // Launch kernel here
-      //dim3 dimBlock(BLOCK_SIZE *BLOCK_SIZE);
-      //dim3 dimGrid((a.rows() / dimBlock.x) * (a.cols() / dimBlock.y));
-      CudaMVKernel<<<m, 256>>>(d_a, d_x, d_y, m, k);
+      dim3 dimBlock(BLOCK_SIZE);
+      dim3 dimGrid(std::ceil((float)m / (BLOCK_SIZE)));
```
Comment (yiskylee) [inline cpp/src/cuda/cuda_matrix_vector_multiply.cu]: A conventional way of set up the grid size without using ceiling function would be:

(m + BLOCK_SIZE -1) / BLOCK_SIZE

Try some examples of m and see if this gives you the same result as using std::ceil()

File: cpp/test/gpu_operations_test/cuda_matrix_vector_multiply_shared_memory_test.cc [hunk 29]
```diff
@@ -90,7 +90,7 @@ TYPED_TEST_CASE(CudaSharedMVMultiplyTest, dataTypes);
 
 TYPED_TEST(CudaSharedMVMultiplyTest, GlobalMemTest) {
   // Create test data
-  int m = 1000;
+  int m = 60000;
```
Comment (yiskylee) [inline cpp/test/gpu_operations_test/cuda_matrix_vector_multiply_shared_memory_test.cc]: So does it pass the GlobalMemTest? I would try other BLOCK_SIZE as well.

File: cpp/include/logistic_regression.h [hunk 30]
```diff
@@ -63,23 +63,18 @@ class LogisticRegression {
   ///
   /// \param input
   /// A Vector containing the theta to manually set the model
-  void setTheta(const Vector<T> &input) {
+  void SetTheta(const Vector<T> &input) {
     theta = input;
```
Comment (yiskylee) [inline cpp/include/logistic_regression.h]: theta_

File: cpp/test/model_test/logistic_regression_test.cc [hunk 31]
```diff
@@ -76,8 +76,9 @@ TYPED_TEST(LogisticRegressionTest, MatrixLogisticRegressionOneModel) {
           6.922, 1.771,
           8.675, -0.242,
           7.673, 3.508;
-  this->testModel1.Fit(this->training_x, this->training_y, this->predict_x, this->iterations,
-        this->alpha);
+  this->testModel1.SetIterations(this->iterations);
```
Comment (yiskylee) [inline cpp/test/model_test/logistic_regression_test.cc]: test_model?

File: cpp/test/model_test/gpu_logistic_regression_test.cc [hunk 32]
```diff
@@ -45,10 +45,8 @@ class GpuLogisticRegressionTest: public ::testing::Test {
   Nice::GpuLogisticRegression<T> gpu_model_;
 
   // Loads data from file
-  Nice::Matrix<T> Filler(std::string fileName, std::string d) {
+  Nice::Matrix<T> Filler(std::string fileName, std::string d, std::string testName) {
```
Comment (yiskylee) [inline cpp/test/model_test/gpu_logistic_regression_test.cc]: variable naming: file_name, test_name

**Review by yiskylee — COMMENTED** [review]
1. Please merge the current develop branch to yours as I have recently fixed some issues in the develop branch.
2. Do "make check" in build directory, and fix coding style errors.
